### DM-06 Practice 1

#### Import libraries 

In [23]:
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.feature_extraction.text import CountVectorizer
from sklearn.naive_bayes import MultinomialNB

#### Parameters  

In [24]:
csv_in = 'newsgroups.csv'

# To show all columns and rows  
pd.options.display.max_columns=999
pd.options.display.max_rows=999

#### Read CSV file  

In [25]:
df = pd.read_csv(csv_in, delimiter=',', skiprows=0, header=0,
                 encoding='latin-1')
print(df.shape)
print(df.info())
display(df.head())

(297, 2)
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 297 entries, 0 to 296
Data columns (total 2 columns):
 #   Column    Non-Null Count  Dtype 
---  ------    --------------  ----- 
 0   category  297 non-null    object
 1   text      297 non-null    object
dtypes: object(2)
memory usage: 4.8+ KB
None


category                                               text
0  baseball    The Orioles' pitching staff again is having ...
1  baseball    I agree and disagree.  John is saying that t...
2  baseball   Hell, the Orioles' Opening Day game could eas...
3  baseball   There's a lot of whining about how much playe...
4  baseball    I doubt Henderson would clear waivers.  And ...

#### Check #texts for each group  

In [26]:
print(df['category'].value_counts())

baseball       99
space          99
electronics    99
Name: category, dtype: int64


#### Obtain X (data) and y (true label)  

In [27]:
X = df['text']
y = df['category']

#### Divide data into train data and test data  

In [28]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.25,
                                                    random_state=5)
print(X_train.shape, y_train.shape)
print(X_test.shape, y_test.shape)

(222,) (222,)
(75,) (75,)


**Ans. 222**  

#### Count words  

In [29]:
vectorizer = CountVectorizer()
vectorizer.fit(X_train)
vocab = vectorizer.get_feature_names()
print('Vocabulary size:', len(vocab))
print(vocab[:10])  # debug

Vocabulary size: 8061
['00', '000', '000000', '0020', '0028', '005', '0065', '01', '02', '02138']


**Ans1. 8061**  

#### Calculate BoW  

In [30]:
X_train_bow = vectorizer.transform(X_train)
X_test_bow = vectorizer.transform(X_test)
print('X_train_bow:')
print(repr(X_train_bow))
print('X_test_bow:')
print(repr(X_test_bow))

X_train_bow:
<222x8061 sparse matrix of type '<class 'numpy.int64'>'
	with 23599 stored elements in Compressed Sparse Row format>
X_test_bow:
<75x8061 sparse matrix of type '<class 'numpy.int64'>'
	with 6858 stored elements in Compressed Sparse Row format>


#### Training of naive Bayes classifier  

In [31]:
model = MultinomialNB(alpha=1.0)
model.fit(X_train_bow, y_train)
print(model.classes_)

['baseball' 'electronics' 'space']


#### Prediction accuracy for training data   

In [32]:
train_score = model.score(X_train_bow, y_train)
print('Train accuracy:', train_score)

Train accuracy: 0.9864864864864865


**Ans2. 98.6%**   

#### Prediction using naive Bayes classifier after training  

In [33]:
# Obtain each prediction  
y_test_pred = model.predict(X_test_bow)
df_pred = pd.DataFrame({
            'pred': y_test_pred,
            'true': y_test
          }).reset_index(drop=True)
display(df_pred.head())

pred         true
0     baseball     baseball
1     baseball  electronics
2  electronics  electronics
3     baseball     baseball
4        space        space

In [34]:
# Make crosstable  
ctab = pd.crosstab(df_pred['pred'], df_pred['true'])
display(ctab)

true         baseball  electronics  space
pred                                     
baseball           22            2      2
electronics         0           17      1
space               0            5     26

In [35]:
# Prediction accuracy  
test_score = model.score(X_test_bow, y_test)
print('Test accuracy:', test_score)

Test accuracy: 0.8666666666666667


**Ans3. 86.7%**  